### ISSUE TITLE
'wwctl server restart' and systemd service are disconnected

### ISSUE THREAD

**Issue by tcooper** [hpcng/warewulf#160]
This is possibly a regression of #37 as I see inconsistent behavior when using `systemd` and `wwctl` to control the warewulf daemon process.

My automation starts `warewulfd` using the `systemd` unit file early on with the method described in the [EL8 Quickstart Guide](https://warewulf.hpcng.org/docs/quickstart/el8)...

### sample vagrant output
```
...
==> mgr: Running provisioner: Warewulf v4 Configuration (shell)...
    mgr: Running: script: Warewulf v4 Configuration
    mgr: + sudo cp /tmp/warewulf.conf /etc/warewulf/warewulf.conf
    mgr: + sudo systemctl restart firewalld
    mgr: + sudo firewall-cmd --permanent --add-service warewulf
    mgr: success
    mgr: + sudo firewall-cmd --permanent --add-service nfs
    mgr: success
    mgr: + sudo firewall-cmd --permanent --add-service tftp
    mgr: success
    mgr: + sudo firewall-cmd --reload
    mgr: success
    mgr: + sudo groupadd -r warewulf
    mgr: + sudo systemctl daemon-reload
    mgr: + sudo systemctl enable --now warewulfd
    mgr: Created symlink /etc/systemd/system/multi-user.target.wants/warewulfd.service → /usr/lib/systemd/system/warewulfd.service.
    mgr: + sudo wwctl server status
    mgr: Warewulf server is running at PID: 32339
...
```

...which works fine

Later, after addition/modification of container, nodes and overlays the the `wwctl server restart` command is executed to verify both mechanisms operate leading to errors and/or failures to control the warewulf daemon process.

### sample vagrant output
```
...
    mgr: + sudo wwctl overlay build runtime default
    mgr: [INFO]     Updating Runtime Overlays...
    mgr: [INFO]     cn0                                : Done
    mgr: [INFO]     cn1                                : Done
    mgr: + sudo wwctl overlay build system default
    mgr: [INFO]     Updating System Overlays...
    mgr: [INFO]     cn0                                : Done
    mgr: [INFO]     cn1                                : Done
    mgr: + sudo wwctl node ready
    mgr: NODE NAME                 STATUS     VNFS   KERNEL KMODS  SYS-OL RUN-OL
    mgr: cn0                       true       true   true   true   true   true
    mgr: cn1                       true       true   true   true   true   true
    mgr: + sudo wwctl server restart
    mgr: Terminated Warewulf server at PID: 32339
    mgr: Started Warewulf server at PID: 32747
    mgr: + sudo wwctl server status
    mgr: ERROR: Warewulf server is not running
The SSH command responded with a non-zero exit status. Vagrant
assumes that this means the command failed. The output for this command
should be in the log above. Please read the output to determine what
went wrong.
```

Shelling into the running VM to inspect reveals...

### warewulfd is not running
```
~/bo/warewulf-vagrant on development ❯ vagrant ssh mgr
[vagrant@mgr ~]$ sudo wwctl server status
ERROR: Warewulf server is not running
```

### manual restart of warewulfd using wwctl fails
```
[vagrant@mgr ~]$ sudo wwctl server restart
Warewulf daemon process not running
Started Warewulf server at PID: 32867

[vagrant@mgr ~]$ sudo wwctl server status
ERROR: failed to send process SIGCONT: os: process already finished
```

### systemd doesn't know about proper PID
```
[vagrant@mgr ~]$ systemctl status warewulfd.service
● warewulfd.service - Warewulf cluster management daemon
   Loaded: loaded (/usr/lib/systemd/system/warewulfd.service; enabled; vendor preset: disabled)
   Active: failed (Result: exit-code) since Thu 2021-09-30 17:07:47 UTC; 3min 9s ago
     Docs: https://hpcng.github.io/warewulf
  Process: 32813 ExecStart=/usr/bin/wwctl server start (code=exited, status=0/SUCCESS)
 Main PID: 32818 (code=exited, status=255)
 
[vagrant@mgr ~]$ cat /var/run/warewulfd.pid
32867
```

### warewulfd.log show multiple failures with processes attempting to open in-use ports
```
[vagrant@mgr ~]$ ls -l /var/log/warewulfd.log
-rw-r--r--. 1 root root 2355 Sep 30 17:09 /var/log/warewulfd.log

[vagrant@mgr ~]$ cat /var/log/warewulfd.log
[Thu Sep 30 17:07:03 UTC 2021] (re)Loading the node Database
[Thu Sep 30 17:07:03 UTC 2021] Starting HTTPD REST service on port 9873
[Thu Sep 30 17:07:43 UTC 2021] (re)Loading the node Database
[Thu Sep 30 17:07:44 UTC 2021] (re)Loading the node Database
[Thu Sep 30 17:07:44 UTC 2021] (re)Loading the node Database
[Thu Sep 30 17:07:44 UTC 2021] (re)Loading the node Database
[Thu Sep 30 17:07:44 UTC 2021] (re)Loading the node Database
[Thu Sep 30 17:07:44 UTC 2021] (re)Loading the node Database
[Thu Sep 30 17:07:44 UTC 2021] (re)Loading the node Database
[Thu Sep 30 17:07:45 UTC 2021] (re)Loading the node Database
[Thu Sep 30 17:07:45 UTC 2021] Starting HTTPD REST service on port 9873
[Thu Sep 30 17:07:45 UTC 2021] (re)Loading the node Database
[Thu Sep 30 17:07:45 UTC 2021] Starting HTTPD REST service on port 9873
ERROR: failed to start Warewulf server: failed to run server: Could not start listening service: listen tcp :9873: bind: address already in use
[Thu Sep 30 17:07:46 UTC 2021] (re)Loading the node Database
[Thu Sep 30 17:07:46 UTC 2021] Starting HTTPD REST service on port 9873
ERROR: failed to start Warewulf server: failed to run server: Could not start listening service: listen tcp :9873: bind: address already in use
[Thu Sep 30 17:07:46 UTC 2021] (re)Loading the node Database
[Thu Sep 30 17:07:46 UTC 2021] Starting HTTPD REST service on port 9873
ERROR: failed to start Warewulf server: failed to run server: Could not start listening service: listen tcp :9873: bind: address already in use
[Thu Sep 30 17:07:46 UTC 2021] (re)Loading the node Database
[Thu Sep 30 17:07:46 UTC 2021] Starting HTTPD REST service on port 9873
ERROR: failed to start Warewulf server: failed to run server: Could not start listening service: listen tcp :9873: bind: address already in use
[Thu Sep 30 17:07:46 UTC 2021] (re)Loading the node Database
[Thu Sep 30 17:07:46 UTC 2021] Starting HTTPD REST service on port 9873
ERROR: failed to start Warewulf server: failed to run server: Could not start listening service: listen tcp :9873: bind: address already in use
[Thu Sep 30 17:09:05 UTC 2021] (re)Loading the node Database
[Thu Sep 30 17:09:05 UTC 2021] Starting HTTPD REST service on port 9873
ERROR: failed to start Warewulf server: failed to run server: Could not start listening service: listen tcp :9873: bind: address already in use
```

Continued use of both `systemd` and `wwctl` to manage the start and stop of the `warewulfd` clearly shows the two mechanisms cannot be used interchangeably. It's possible this is limited to the `wwctl server restart` functionality but that is not clear.

```
[vagrant@mgr ~]$ sudo systemctl status warewulfd.service
● warewulfd.service - Warewulf cluster management daemon
   Loaded: loaded (/usr/lib/systemd/system/warewulfd.service; enabled; vendor preset: disabled)
   Active: inactive (dead) since Thu 2021-09-30 17:35:42 UTC; 3min 59s ago
     Docs: https://hpcng.github.io/warewulf
  Process: 33280 ExecStart=/usr/bin/wwctl server start (code=exited, status=0/SUCCESS)
 Main PID: 33286 (code=killed, signal=TERM)

Sep 30 17:35:04 mgr.example.com systemd[1]: Starting Warewulf cluster management daemon...
Sep 30 17:35:04 mgr.example.com wwctl[33280]: Started Warewulf server at PID: 33286
Sep 30 17:35:04 mgr.example.com systemd[1]: Started Warewulf cluster management daemon.
Sep 30 17:35:42 mgr.example.com systemd[1]: Stopping Warewulf cluster management daemon...
Sep 30 17:35:42 mgr.example.com systemd[1]: warewulfd.service: Succeeded.
Sep 30 17:35:42 mgr.example.com systemd[1]: Stopped Warewulf cluster management daemon.

[vagrant@mgr ~]$ sudo systemctl start warewulfd.service

[vagrant@mgr ~]$ sudo systemctl status warewulfd.service
● warewulfd.service - Warewulf cluster management daemon
   Loaded: loaded (/usr/lib/systemd/system/warewulfd.service; enabled; vendor preset: disabled)
   Active: active (running) since Thu 2021-09-30 17:40:27 UTC; 1s ago
     Docs: https://hpcng.github.io/warewulf
  Process: 33388 ExecStart=/usr/bin/wwctl server start (code=exited, status=0/SUCCESS)
 Main PID: 33393 (wwctl)
    Tasks: 7 (limit: 12438)
   Memory: 8.2M
   CGroup: /system.slice/warewulfd.service
           └─33393 /usr/bin/wwctl server start

Sep 30 17:40:27 mgr.example.com systemd[1]: Starting Warewulf cluster management daemon...
Sep 30 17:40:27 mgr.example.com wwctl[33388]: Started Warewulf server at PID: 33393
Sep 30 17:40:27 mgr.example.com systemd[1]: Started Warewulf cluster management daemon.

[vagrant@mgr ~]$ cat /var/log/warewulfd.log
[Thu Sep 30 17:40:27 UTC 2021] (re)Loading the node Database
[Thu Sep 30 17:40:27 UTC 2021] Starting HTTPD REST service on port 9873

[vagrant@mgr ~]$ cat /var/run/warewulfd.pid
33393

[vagrant@mgr ~]$ sudo wwctl server status
Warewulf server is running at PID: 33393

[vagrant@mgr ~]$ sudo wwctl server restart
Terminated Warewulf server at PID: 33393
Started Warewulf server at PID: 33424

[vagrant@mgr ~]$ sudo wwctl server status
ERROR: Warewulf server is not running

[vagrant@mgr ~]$ cat /var/run/warewulfd.pid
cat: /var/run/warewulfd.pid: No such file or directory

[vagrant@mgr ~]$ sudo systemctl status warewulfd.service
● warewulfd.service - Warewulf cluster management daemon
   Loaded: loaded (/usr/lib/systemd/system/warewulfd.service; enabled; vendor preset: disabled)
   Active: failed (Result: exit-code) since Thu 2021-09-30 17:41:10 UTC; 1min 3s ago
     Docs: https://hpcng.github.io/warewulf
  Process: 33486 ExecStart=/usr/bin/wwctl server start (code=exited, status=0/SUCCESS)
 Main PID: 33491 (code=exited, status=255)

Sep 30 17:41:10 mgr.example.com systemd[1]: warewulfd.service: Service RestartSec=100ms expired, scheduling restart.
Sep 30 17:41:10 mgr.example.com systemd[1]: warewulfd.service: Scheduled restart job, restart counter is at 6.
Sep 30 17:41:10 mgr.example.com systemd[1]: Stopped Warewulf cluster management daemon.
Sep 30 17:41:10 mgr.example.com systemd[1]: warewulfd.service: Start request repeated too quickly.
Sep 30 17:41:10 mgr.example.com systemd[1]: warewulfd.service: Failed with result 'exit-code'.
Sep 30 17:41:10 mgr.example.com systemd[1]: Failed to start Warewulf cluster management daemon.

[vagrant@mgr ~]$ sudo wwctl server status
ERROR: Warewulf server is not running
```

At this point manual intervention is required to kill the running wwctl process that is holding open the LISTEN socket...

```
[vagrant@mgr ~]$ pgrep -a -f wwctl
33424 wwctl server start

[vagrant@mgr ~]$ sudo lsof -p 33424
COMMAND   PID USER   FD      TYPE DEVICE SIZE/OFF      NODE NAME
wwctl   33424 root  cwd       DIR    8,1      135 101148691 /home/vagrant
wwctl   33424 root  rtd       DIR    8,1      240       128 /
wwctl   33424 root  txt       REG    8,1 26906800 102419642 /usr/bin/wwctl
wwctl   33424 root  mem       REG    8,1    95456      9784 /usr/lib64/libz.so.1.2.11
wwctl   33424 root  mem       REG    8,1  3071432     20638 /usr/lib64/libcrypto.so.1.1.1g
wwctl   33424 root  mem       REG    8,1  3168120      9645 /usr/lib64/libc-2.28.so
wwctl   33424 root  mem       REG    8,1    28968      9647 /usr/lib64/libdl-2.28.so
wwctl   33424 root  mem       REG    8,1   320504      9659 /usr/lib64/libpthread-2.28.so
wwctl   33424 root  mem       REG    8,1   278432      9638 /usr/lib64/ld-2.28.so
wwctl   33424 root    0r      CHR    1,3      0t0      9273 /dev/null
wwctl   33424 root    1u      REG    8,1     1651 102478881 /var/log/warewulfd.log
wwctl   33424 root    2u      REG    8,1     1651 102478881 /var/log/warewulfd.log
wwctl   33424 root    3u     IPv6  61685      0t0       TCP *:9873 (LISTEN)
wwctl   33424 root    4u  a_inode   0,14        0     10263 [eventpoll]
wwctl   33424 root    5r     FIFO   0,13      0t0     61680 pipe
wwctl   33424 root    6w     FIFO   0,13      0t0     61680 pipe
```

**Comment by gmkurtzer** [comment]
Hi @tcooper, what version are you using? I think I fixed this in main a few weeks ago.

```
[root@localhost ~]# systemctl start warewulfd.service 
[root@localhost ~]# wwctl server status
Warewulf server is running at PID: 221314
```

**Comment by tcooper** [comment]
Does `wwctl` even have a `version` command or `--version` flag?

```
[vagrant@mgr ~]$ sudo wwctl version
ERROR: unknown command "version" for "wwctl"

[vagrant@mgr ~]$ sudo wwctl --version
ERROR: unknown flag: --version
```

Seems like no.

I guess you have to trust the one I'm running came from here...

```
[vagrant@mgr ~]$ cd warewulf/
[vagrant@mgr warewulf]$ git status
On branch main
Your branch is up to date with 'origin/main'.

nothing to commit, working tree clean

[vagrant@mgr warewulf]$ git remote -v
origin	https://github.com/hpcng/warewulf.git (fetch)
origin	https://github.com/hpcng/warewulf.git (push)

[vagrant@mgr warewulf]$ git --no-pager log -n1
commit 1d48c9ce15d68c04af0079e2b290b3223befad1d (HEAD -> main, origin/main, origin/HEAD)
Author: Gregory Kurtzer <gmkurtzer@gmail.com>
Date:   Wed Sep 29 22:41:58 2021 -0700

    Removed static port delcaration in the dhcpd templates
```

**Comment by tcooper** [comment]
And... I just pulled a fresh copy or the repo, cleaned, built, installed to get...

### here is version information
```
[vagrant@mgr ~]$ cd warewulf

[vagrant@mgr warewulf]$ git --no-pager log -n1
commit bbf6398fed7035be7b22236bcfbd6558562aa135 (HEAD -> main, origin/main, origin/HEAD)
Merge: 604dd20 cb1455b
Author: Gregory M. Kurtzer <gmkurtzer@gmail.com>
Date:   Fri Oct 1 00:09:05 2021 -0700

    Merge pull request #162 from gmkurtzer/profile_discoverable

    Allow `--discoverable` option to be set via profiles
```

### systemd sees warewulfd running
```
[vagrant@mgr ~]$ systemctl status warewulfd.service
● warewulfd.service - Warewulf cluster management daemon
   Loaded: loaded (/usr/lib/systemd/system/warewulfd.service; enabled; vendor preset: disabled)
   Active: active (running) since Fri 2021-10-01 19:31:25 UTC; 1min 48s ago
     Docs: https://warewulf.hpcng.org/
  Process: 42989 ExecStart=/usr/bin/wwctl server start (code=exited, status=0/SUCCESS)
 Main PID: 42994 (wwctl)
    Tasks: 7 (limit: 12438)
   Memory: 17.4M
   CGroup: /system.slice/warewulfd.service
           └─42994 /usr/bin/wwctl server start
```

### wwctl works fine
```
[vagrant@mgr ~]$ sudo wwctl node list
NODE NAME              PROFILES                   NETWORK
================================================================================
cn0                    default                    enp0s8:192.168.15.130
cn1                    default                    enp0s8:192.168.15.131
```


### wwclt sees daemon running
```
[vagrant@mgr ~]$ sudo wwctl server status
Warewulf server is running at PID: 42994

[vagrant@mgr ~]$ cat /var/run/warewulfd.pid
42994
```

### attempt restart of warewulf daemon using wwctl
```
[vagrant@mgr ~]$ sudo wwctl server restart
Terminated Warewulf server at PID: 42994
Started Warewulf server at PID: 43108
```

### PID file is gone
```
[vagrant@mgr ~]$ cat /var/run/warewulfd.pid
cat: /var/run/warewulfd.pid: No such file or directory
```


### wwctl does not see daemon running
```
[vagrant@mgr ~]$ sudo wwctl server status
ERROR: Warewulf server is not running
```

### systemd does not see daemon running
```
[vagrant@mgr ~]$ systemctl status warewulfd.service
● warewulfd.service - Warewulf cluster management daemon
   Loaded: loaded (/usr/lib/systemd/system/warewulfd.service; enabled; vendor preset: disabled)
   Active: failed (Result: exit-code) since Fri 2021-10-01 19:33:42 UTC; 29s ago
     Docs: https://warewulf.hpcng.org/
  Process: 43166 ExecStart=/usr/bin/wwctl server start (code=exited, status=0/SUCCESS)
 Main PID: 43171 (code=exited, status=255)
```

### There is a deamon running
```
[vagrant@mgr ~]$ pgrep -a -f wwctl
43108 wwctl server start
```

*NOTE: The PID shown here is the same PID as printed by `wwctl server restart`.*

This uncontrolled running daemon will prevent any new stop/start actions from succeeding until it is killed manually.

**Comment by gmkurtzer** [comment]
Let's document this as a known issue in the 4.2.0 release notes. (from @tcooper)

**Comment by mslacken** [comment]
Seems to be fixed in 4.30. Perhaps #550 fixed this.
